docs/DEPLOYMENT.md

```diff
@@ -0,0 +1,195 @@
+workers, and tracking and reporting their progress and results via a visual
+dashboard (caveat: many of these pieces are WIP as r0 of this doc).
+
+The above entails managing infrastructure and environments. To automate such
+tasks, we've chosen a synergistic combination of popular tools:
+
+* *Terraform configurations* to manage the physical infrastructure. These are
+  _cloud-specific_. The PL team runs testground on AWS, and we've open sourced
+  our configurations.
+    * Users on AWS can borrow these configs and take off. Users using other
+      providers will need to replace the Terraform configs.
+* *Ansible playbooks* to manage the logical infrastructure, such as Docker
+  services, system services, overlay networks, etc. The Ansible scripts are
+  _cloud-leaky_. In theory they could be cloud-agnostic, but in practice various
+  items such as S3 storage, AWS ECR authentication, etc. are cloud-specific.
+    * Users wishing to deploy testground on an alternate provider, will need to
+      patch those elements.
+* _[terraform-inventory](https://github.com/adammck/terraform-inventory)_ for
+  the hand-off between the two.
+
+These assets handle the intantiation of various node types of nodes, and the
+backing infrastructural elements (e.g. machines, VPCs, security groups, etc.):
+
+1. **Testground master**: runs the daemon, watches GitHub, exposes an
+   authenticated API over HTTPS, serves the dashboard, keeps track of run state,
+   schedules incoming runs, governs and scales the worker pool size.
+2. **Testground worker**: runs testground workloads.
+3. **Testground redis**: a dedicated instance to run the sync service backend
+   (Redis).
+
+We think of the deployment generated by these as a _prototype deployment_. We
+run multiple replicas of this prototype as specified in the next section.
+
+### Environments
+
+At Protocol Labs, we run multiple deployments of testground to cater for the
+IPFS and libp2p projects. We call these "environments".
+
+An environment belongs to one of four classes. Each environment is identified by
+a unique name, and this name is used in all its resources for scoping and
+identification purposes.
+
+1. **Isolated developer playground:** dedicated entirely to a single developer;
+   this is their laboratory. Created and torn down at will. They can make or
+   break anything they want here. It will not affect test runs in other
+   environments.
+
+     * URL: `dev-<name>.testground.ipfs.team`
+
+     * Availability: Developers don't work 24/7, and neither should their
+       playground be powered on 24/7. Doing so would be wasteful. At the same
+       time, we don't want to destroy and recreate the environment frequently;
+       that takes time, and there's nothing worse than sitting with your arms
+       crossed / context switching for 10 minutes before you can start working.
+
+     * Scaling: the scale is determined manually. The developer uses
```

Comment:
Who would be the user of this? Let's survey developers if this is a workflow that they feel comfortable with, because I am not convinced everyone is going to be happy setting up infrastructure by themselves. This is rather time consuming and takes focus always from application development.

Reply:
👍